// Copyright 2017 Intel Corporation
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// -----------------------------------------------------------------------------

syntax = "proto3";
import "application.proto";

option java_multiple_files = true;
option java_package = "sawtooth.supplychain.protobuf";

// Supplychain Transaction Setting Payload
message SupplyChainPayload {
    // The action indicates data is contained within this payload
    enum Action {
        AGENT_CREATE = 0;
        APPLICATION_CREATE = 1;
        APPLICATION_ACCEPT = 2;
        APPLICATION_REJECT = 3;
        APPLICATION_CANCEL = 4;
        RECORD_CREATE = 5;
        RECORD_FINALIZE = 6;
    }

    // The action of this payload
    Action action = 1;

    // The content of this payload
    bytes data = 2;
}

message AgentCreatePayload {
    string name = 1;
}

message ApplicationCreatePayload {
    string record_identifier = 1;
    int64 creation_time = 2;
    Application.Type type = 3;
    string terms = 4;
}

message ApplicationAcceptPayload {
    string record_identifier = 1;
    string applicant = 2;
    Application.Type type = 3;
    int64 timestamp = 4;
}

message ApplicationRejectPayload {
    string record_identifier = 1;
    string applicant = 2;
    Application.Type type = 3;
}

message ApplicationCancelPayload {
    string record_identifier = 1;
    string applicant = 2;
    Application.Type type = 3;
}

message RecordCreatePayload {
    string identifier = 1;
    int64 creation_time = 2;
}

message RecordFinalizePayload {
    string identifier = 1;
}
